File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.js
/* /dev/stdin */
let fs = require("fs");
let input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .split("\n")
  .map((x) => x.replaceAll("\r", ""));

const N = Number(input[0]);
const myCards = input[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

const M = Number(input[2]);
const targetCards = input[3].split(" ").map(Number);
const answer = [];

for (const targetCard of targetCards) {
  const lower = lowerBound(myCards, targetCard);
  const upper = upperBound(myCards, targetCard);

  const result = upper - lower;
  answer.push(result);
}
console.log(answer.join(" "));

function lowerBound(myCards, targetCard) {
  let left = 0;
  let right = myCards.length;

  while (left < right) {
    const mid = Math.floor((left + right) / 2);

    if (myCards[mid] >= targetCard) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }

  return left;
}

function upperBound(myCards, targetCard) {
  let left = 0;
  let right = myCards.length;

  while (left < right) {
    const mid = Math.floor((left + right) / 2);

    if (myCards[mid] > targetCard) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }

  return left;
}

File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.py
# 입력 받기
n = int(input())
cards = list(map(int, input().split()))
m = int(input())
targets = list(map(int, input().split()))

# 각 숫자가 몇 개씩 있는지 저장할 딕셔너리 초기화
card_count = {}

# 카드 숫자들을 세기
for card in cards:
    if card in card_count:
        card_count[card] += 1
    else:
        card_count[card] = 1

# 결과 출력
result = []
for target in targets:
    if target in card_count:
        result.append(card_count[target])
    else:
        result.append(0)

# 결과 출력
print(' '.join(map(str, result)))
